### PR TITLE
Correct `list.last` bug due to ref and on clause, add tests

### LINKED ISSUE #15366: List standard module `last()` method modifies the list.
### Summary of Problem

Calling the list `last` method returns a reference to the first element and assigns the value of the last element to it.

### Steps to Reproduce

**Source Code:**

```chapel
use List;
use List;
var l = new list(int);
l.append(1);
l.append(2);
ref x = l.last();
writeln(x);
writeln(l);// prints [2, 2]
x = 6;
writeln(l); // prints [6, 2]
```

**Compile command:**
`chpl list.chpl`

**Execution command:**
`./list`

### Configuration Information

- Output of `chpl --version`: chpl version 1.21.0 pre-release (2f36c637a4)
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_LLVM: none
CHPL_AUX_FILESYS: none
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR fixes a rather nasty bug in `list.last` caused by a combination of `on` clauses and my misunderstanding of Chapel references (they can't be reassigned!). It also adds tests to lock in the correct behavior for `list.last` and `list.first` (which also exhibited the bug).

Resolves #15366.

---

Testing:

- [x] `ALL` on `linux64` when `CHPL_COMM=none`
- [x] `library/standard/List/listRefBugMultiLoc.chpl` on `darwin` when `CHPL_COMM=gasnet`

### CODE REVIEW

**Comment by bradcray** [comment]
Knowing that the on-clauses have changed here, I'd recommend writing a test that calls and assigns to first and last from a locale other than where the list is stored to make sure it works.  That is, something like:

```chapel
var l: list(int);
l.append(1);
l.append(2);
on Locales[numLocales-1] {
  l.first() = 3;
  l.last() = 4;
  writeln(l);
}
writeln(l);
```

with a `.numlocales` file setting it to 2 for gasnet runs.

**Comment by dlongnecke-cray** [comment]
Sure, I can do that.

File: test/library/standard/List/listRefBugMultiLoc.skipif [hunk 1]
```diff
@@ -0,0 +1 @@
+CHPL_COMM==none
```
Comment (bradcray) [inline test/library/standard/List/listRefBugMultiLoc.skipif]: You could drop this skipif since the test should still be valid and work on a single locale.

Comment (dlongnecke-cray) [reply test/library/standard/List/listRefBugMultiLoc.skipif]: Sounds good.

**Pull request merged**